klayout/drc/rule_decks/main.drc

```diff
@@ -0,0 +1,1072 @@
+
+ngate_dn        = ngate.and(dnwell_p)
+ptap_dn         = ptap.and(dnwell_p).outside(well_diode_mk)
+
+pgate_dn        = pgate.and(dnwell_n)
+ntap_dn         = ntap.and(dnwell_n)
+
+psd_dn          = pcomp.and(dnwell_n).interacting(pgate_dn).not(pgate_dn).not(res_mk)
+nsd_dn          = ncomp.and(dnwell_p).interacting(ngate_dn).not(ngate_dn).not(res_mk)
+
+natcomp        	= nat.and(comp)
+
+# Gate
+nom_gate = tgate.not(dualgate)
+thick_gate = tgate.and(dualgate)
+
+ngate_56V = ngate.and(dualgate)
+pgate_56V = pgate.and(dualgate)
+
+ngate_5V = ngate_56V.and(v5_xtor)
+pgate_5V = pgate_56V.and(v5_xtor)
+
+ngate_6V = ngate_56V.not(v5_xtor)
+pgate_6V = pgate_56V.not(v5_xtor)
+
+# DNWELL
+dnwell_3p3v = dnwell.not_interacting(v5_xtor).not_interacting(dualgate)
+dnwell_56v = dnwell.overlapping(dualgate)
+
+# LVPWELL
+lvpwell_dn = lvpwell.interacting(dnwell)
+lvpwell_out = lvpwell.not_interacting(dnwell)
+
+lvpwell_dn3p3v = lvpwell.and(dnwell_3p3v)
+lvpwell_dn56v = lvpwell.and(dnwell_56v)
+
+# NWELL
+nwell_dn = nwell.interacting(dnwell)
+nwell_n_dn = nwell.not_interacting(dnwell)
+
+#================================================
+#------------- LAYERS CONNECTIONS ---------------
+#================================================
+
+if CONNECTIVITY_RULES
+
+  logger.info("Construct connectivity for the design.")
+
+  connect(dnwell,  ncomp)
+  connect(ncomp,  contact)
+  connect(pcomp,  contact)
+
+  connect(lvpwell_out,  pcomp)
+  connect(lvpwell_dn, pcomp)
+
+  connect(nwell,  ncomp)
+  connect(natcomp,  contact)
+  connect(mvsd,  ncomp)
+  connect(mvpsd,  pcomp)
+  connect(contact,  metal1)
+  connect(metal1,  via1)
+  connect(via1,    metal2)
+  if METAL_LEVEL != "2LM"
+    connect(metal2,  via2)
+    connect(via2,    metal3)
+
+    if METAL_LEVEL != "3LM"
+      connect(metal3,  via3)
+      connect(via3,    metal4)
+      
+      if METAL_LEVEL != "4LM"
+        connect(metal4,  via4)
+        connect(via4,    metal5)
+
+        if METAL_LEVEL != "5LM"
+          connect(metal5,  via5)
+          connect(via5,    metaltop)
+        end
+      end
+    end
+  end
+
+end #CONNECTIVITY_RULES
+
+#================================================
+#------------ PRE-DEFINED FUNCTIONS -------------
+#================================================
+
+def conn_space(layer,conn_val,not_conn_val, mode)
+  if conn_val > not_conn_val
+    raise "ERROR : Wrong connectivity implementation"
+  end
+  connected_output = layer.space(conn_val.um, mode).polygons(0.001)
+  unconnected_errors_unfiltered = layer.space(not_conn_val.um, mode)
+  singularity_errors = layer.space(0.001.um)
+  # Filter out the errors arising from the same net
+  unconnected_errors = DRC::DRCLayer::new(self, RBA::EdgePairs::new)
+  unconnected_errors_unfiltered.data.each do |ep|
+    net1 = l2n_data.probe_net(layer.data, ep.first.p1)
+    net2 = l2n_data.probe_net(layer.data, ep.second.p1)
+    if !net1 || !net2
+      puts "Should not happen ..."
```

Comment:
assert + error?

Reply:
Connectivity based checks will change according to a discussion that I got from @klayoutmatthias 

Raising an exception here for that might stop a long running DRC run. Logging and continuing would be definitely better.

Reply:
is there a log error message to use, should we use the logger with the appropriate level instead of `puts`?

Reply:
Done